Exploring nested sampling

fehiepsi opened the pull request:
### Motivation

+ According to #643 and this [forum topic](https://forum.pyro.ai/t/fitting-a-model-with-a-multimodal-posterior-using-flows-and-neutra-hmc/2297/4), it seems that nested sampling is a pretty popular sampler in astrology, which can capture multi-modal posteriors.
+ @Joshuaalbert is developing nested sampling at https://github.com/Joshuaalbert/jaxns and is interested in an integration with NumPyro.
+ Currently, other nested sampling packages require defining `log_likelihood` and `transforms`. I hope that with NumPyro, setting up a model to do nested sampling is more straightforward. So users won't have to know how to construct `log_likelihood` or `transforms` from a model.

### TODO
- [x] Implement UniformReparam which leverages transformations to `icdf` method
- [x] Add tests with/without enum
- [x] Add gaussian shells example
- [x] Add cdf/icdf methods to some distributions (exercised with current `cdf/icdf` tests)

**Note:** This is just a prototype work, which serves as a reference for future discussions and development.

Blocked by #940 .

Joshuaalbert commented:
Very nice! There is one small change needed. The weight of each sample in nested sampling is not 1, as they are from MCMC. Rather the log_weight of each sample is given by `results.log_p`. You can resample the samples with these weights using `jaxns.utils.resample(key, results.samples, results.log_p)` it will compute the ESS from `log_p` and sample ESS samples with replacement from the samples. Without doing this resampling the summary statistics won't make sense either.

fehiepsi commented:
> The weight of each sample in nested sampling is not 1

Yeah, thanks! It sounds more reasonable now. Looking forward to chat with you in the meeting to understand more about your futuer plan with `jaxns`. Thanks for the great work! :)

Joshuaalbert commented:
Yes, the time you sent works. That email doesn't connect to gmail calendars though so I'll join from a different account.

Joshuaalbert commented:
Here's a good test case: https://github.com/Joshuaalbert/jaxns/blob/master/jaxns/examples/mvn_data_mvn_prior.py
We know exactly what the posterior mean and covariance should be as well as evidence (but you can't compare NUTS to that).

fehiepsi commented:
Thanks, @Joshuaalbert! For testing, we can also use some conjugacy models such as BetaBernoulli... I just updated `jaxns.ipynb` notebook with your mvn test but found that it is tricky to get good results when I changed `data_mu = jnp.ones(ndims)` to `data_mu = jnp.arange(ndims)` while NUTS is pretty fast and just works out of the box. You can try it yourself by changing `data_mu` in that `mvn_data_mvn_prior` example. But there is a good news that with some configurations, the nested sampler can capture the multimodal posterior in the forum topic. :D

Joshuaalbert commented:
@fehiepsi yes, when the prior and likelihood overlap is really small (as it is with data_mu=arange(ndims)) then it can struggle (for now). As I put more development into it, this will get better. I have a few ideas to implement yet that will make it better. I haven't had anyone using besides me, so now that others will use it, I'll have a chance to get feedback like this.

fehiepsi commented:
Thanks, Joshua! It is good to know. I am looking forward to seeing further enhancements from you. The current performance (at least for the forum topic's model) is pretty good in my opinion.

Joshuaalbert commented:
Hi @fehiepsi, I thought we should have a check in. I've been working on two projects that use jaxns, and in doing so found a few edge cases, and made a few upgrades to solve them. I'm soon going to push these, freeze the API, tidy things up, and officially release on pip.

Joshuaalbert commented:
Oh yes, and some other cool features that might be interesting to you. Ability to have discrete RVs, and priors without a defined inverse quantile function.

fehiepsi commented:
Awesome! I am looking forward to the next jaxns release. In the meantime, I'll add some tests and examples to the docstring here. Btw, could you add an instruction for people to cite your work? (I just want to deliver a clear message in the docstring in case people use jaxns wrapper in their research)

Joshuaalbert commented:
yes, regarding citation. I'm going to write a paper. This will take some time, maybe by end January would do.

ChristianMichelsen commented:
> "pretty popular sampler in astrology"

Freudian slip? ;)

Joshuaalbert commented:
By the first version of JAXNS paper is up on arxivs: https://arxiv.org/abs/2012.15286.
I'll release JAXNS on pip soon as I push the new API. It should allow a much broader class of priors, not only quantile transformed ones.

fehiepsi commented:
Hi @Joshuaalbert, I have cleaned up the docs, added some discrete reparams (but I'm not sure if I need to do something else to match the discussion in your paper `JAXNS handles this by having separate prior domain subspaces for priors lacking a transformation from the unit cube, and letting the slice sampler handle these subspaces accordingly.` - could you take a look?), and added a small example on the usage. The recommended default parameters also follow your paper and I took some random choices (based on your examples) for `num_live_points` and `max_samples`. I also did some tests in GPU but it is pretty slow comparing to CPU, probably because the models that I tested on are small and the sampling is sequential.

Joshuaalbert commented:
I think the NS algorithm won't benefit from being run on the GPU, but expensive likelihoods will. The user should then jit-compile the likelihood with a GPU backend and jit-compile the NS smapling on the CPU backend.
But, this should be a per-case decision. E.g. when there are big linear algebra calculations, these can be done on the GPU faster. Also, once I parallelise sampling, maybe this can be tested on GPU. I haven't pushed yet the separate prior subspaces yet, but will do so soon.

fehiepsi commented:
Agree! I thought taking advantage of acceleration hardware is probably a selling point for your work because no other packages support GPU afaik. :)

Joshuaalbert commented:
@fehiepsi JAXNS is now on pip. There's a slightly new API now. Mainly, all non-jitable parameters go into the initialisation of `NestedSampler`, and only jitable parameters, currently `key` and `termination_frac`, are passed the the `__call__` method of the resulting object. Parallelisation of the sampling is almost complete, and I'll push to pip once it's done. New with the latest release is a faster more robust implementation of two sampler choices `'slice'` and `'multi_ellipsoid'`. I also fixed some bugs in the correctness of the implementations, which are now verified correct. Let me know how it goes integrating it.

fehiepsi commented:
Awesome! Thanks for the update, @Joshuaalbert. I'll push forward to have a nested sampling wrapper for NumPyro models in the next NumPyro release.

Joshuaalbert commented:
Hi @fehiepsi I note you're doing some work to incorperate the new API. Let me know if you need pointers.

fehiepsi commented:
@Joshuaalbert The new API looks cleaner to me. But I am a bit concern about the logic [here](https://github.com/pyro-ppl/numpyro/pull/807/files#diff-16765ffeb8303369ba3813350328607783a1068277623873aea27b6ebb2ad5b7R218-R222) where I filtered out `-inf` samples and log_weights. Those `-inf` values do not happen before the last release, so I am wondering if they are expected?

fehiepsi commented on numpyro/contrib/nested_sampling.py:
```diff
@@ -0,0 +1,254 @@
+        :param int num_samples: The number of samples.
+        :return: a dict of posterior samples
+        """
+        p = nn.softmax(self._log_weights)
```
Is there any difference?

Joshuaalbert replied:
I would say the difference is maintainability. If, for example, I were to change the way `samples` and `log_p` were represented, then I would update `resample` and you wouldn't see a difference.

fehiepsi replied:
I see. Currently, it is not straightforward to convert numpyro distributions to jaxns priors so I transformed the model (with some limitations) to a model with only `Uniform(0, 1)` priors. So the results obtained from jaxns nested sampler are samples in unit cube domains and I need to do an extra job converting them back to the original domains. This requires some inspection of jaxns `results` (like overwriting the `samples` field). To be more informed of upstream changes, it is probably better to add some precision tests and pin jaxns version in each numpyro release.

fehiepsi commented on numpyro/contrib/nested_sampling.py:
```diff
@@ -0,0 +1,254 @@
+        log_weights = results.log_p[:num_samples]
+        base_samples = {k: v[:num_samples] for k, v in results.samples.items()}
+        # filter out samples with -inf weights and invalid values
+        valid_masks = [jnp.isfinite(log_weights)] + \
+            [((v > 0) & (v < 1)).reshape((num_samples, -1)).all(-1)
+             for v in base_samples.values()]
+        mask = jnp.stack(valid_masks, -1).all(-1)
+        self._log_weights = log_weights[mask]
```
I saw some `-inf` in log_weights[:num_samples] and in samples[:num_samples] of the example above (in the docstring of NestedSampler), which prevented me from collecting some statistics after sampling. I saw your plotting framework also does some sort of [filtering](https://github.com/Joshuaalbert/jaxns/blob/master/jaxns/plotting.py#L84) after getting `samples[:num_samples]` so I thought that those -inf are expected.

Joshuaalbert replied:
Those lines are related to giving samples with `inf` no weight. I.e. it happens that the you feed in a 1 to the normal distribution quantile function you'll get `inf`. In principle, this shouldn't happen because the samplers try to avoid that happening, but the plotting filters those out for safety. In general, we don't need to filter out logL=-inf for plotting.

`log_weights[:num_samples]` shouldn't have any inf. Make sure you use jaxns=0.0.5 as there would have been some in smaller versions.

fehiepsi replied:
Confirms that those `inf` are gone in jaxns 0.0.5. I'll ping jaxns to `0.0.5` to avoid issues like this.

Joshuaalbert replied:
great

fehiepsi commented on numpyro/contrib/nested_sampling.py:
```diff
@@ -0,0 +1,254 @@
+                               collect_samples=True)
+        results = ns(rng_sampling, termination_frac=self.termination_frac)
+
+        num_samples = int(device_get(results.num_samples))
```
Yes, I thought jittable is not necessary. But it is a good idea to support that. I think I can change the `Predictive` logic to do the transform in a while loop until the counter > results.num_samples. Let me see. Probably it is better to move some print statements to the `diagnostics` method too.

Joshuaalbert replied:
Note, there is the `jaxns.utils.summary` function that does a print summary of the results.

fehiepsi commented on numpyro/contrib/nested_sampling.py:
```diff
@@ -0,0 +1,254 @@
+        self._log_weights = None
+        self._results = None
+
+    def run(self, rng_key, *args, **kwargs):
```
Thanks, I'll update the code to make this jittable.

fehiepsi replied:
> If the user wants to marginalise over the samples, then they should add function to be marginalised over as NestedSampler(..., marginalised=dict(key=callable(**priors), ...)) where each marginalised function has the same call signature as the log-likelihood.

This is nice! I'll add a note in the code to have this functionality in a future PR.

Joshuaalbert replied:
Great!

fehiepsi commented on numpyro/contrib/tfp/distributions.py:
```diff
@@ -149,6 +149,12 @@ def is_discrete(self):
+    def sample_with_intermediates(self, key, sample_shape=()):
+        return self.sample(seed=key, sample_shape=sample_shape), []
+
+    def icdf(self, q):
```
Good idea! But implementing an interface for such approximation can be tricky. I guess we can make something like `TruncatedApproximateReparam` so that users can decide if they want that approximation.

Btw, it seems that the log probability of Gamma there should be
```
return -gammaln(k) - k * jnp.log(theta) + (k - 1.) * jnp.log(X) - X / theta
```

Joshuaalbert replied:
Yes, it could be tricky. If you knew how to estimate the support of the RV, as is done for the gamma example using the interval `(0, mean + 20. * stddev)`, then you could do it quite simply. 

And thanks! Good catch!

fehiepsi commented:
Hi @Joshuaalbert, thanks for reviewing! I just made some changes based on your suggestions. The code looks cleaner now. While trying to add some tests, I faced this issue
```python
from jax import config, numpy as jnp, random
config.update("jax_platform_name", "cpu")
from jaxns.nested_sampling import NestedSampler
from jaxns.prior_transforms import UniformPrior, PriorChain

def log_likelihood(x, **kwargs):
    return jnp.array(0.)

x = UniformPrior('x', 0., 1.)
prior_chain = x.prior_chain()

ns = NestedSampler(loglikelihood=log_likelihood, prior_chain=prior_chain,
                   num_live_points=1000)
ns(random.PRNGKey(32564))
```
It seems to me that the `ns` does not stop with this likelihood. Do you have an intuition for what is the reason?

Joshuaalbert commented:
This was fixed in https://github.com/Joshuaalbert/jaxns/issues/14

Plateaus are where the log-likelihood is the same for different points in the live points. Vanilla nested sampling assumes a strictly monotonic likelihood, but that's not required. Your test now passes. I should probably push this to pip. I'll let you know when I push it.

fehiepsi commented:
Thanks, @Joshuaalbert! I'll take another pass to clean up the wrapper when it is ready.

Joshuaalbert commented:
@fehiepsi I pushed 0.0.6 to pip and it solves this issue.

fehiepsi commented:
Thank you for letting me know, @Joshuaalbert! I was a bit busy recently but will try to update this PR later next week.

fehiepsi commented:
@Joshuaalbert It is nice that the tests are passing now. :+1: Could you pump the package to a newer version? There are some import issues in the last release, which has been fixed in the `jaxns` master branch.

Joshuaalbert commented:
Done. 0.0.7 is out and it's even a little more efficient.

Joshuaalbert commented:
I want to make a note that when you're only trying to do posterior estimation, and want to ignore the evidence calculation, then you can often get away with faster default setting without losing posterior distribution accuracy. The subtle differences that give you an accurate evidence calculation are basically non-noticeable for people only interested in posterior estimation. When you're only interested in posterior estimation, a suitable estimation might be:

``` python
sampler_kwargs=dict(num_slices=prior_chain.Udims*4)
num_live_points=prior_chain.Udims*20
termination_frac=0.3
```

fehiepsi commented:
Thanks, @Joshuaalbert! I'm not sure who are the users of this NestedSampler wrapper, so would like to defer that recommendation to the future. I think all issues that I faced so far have been resolved.

fehiepsi commented:
@fritzo When you have time, could you help me review the design of this sampler, especially the `uniform_reparam_transform` that uses `cdf` methods to transform a distribution over a `support` domain to a uniform distribution over the unit cube? After that transform, we can assume that all latent variables have Uniform(0, 1) prior and perform nested sampling over that transformed model. Because this is a wrapper and lies in `contrib` module, I guess it is fine to not cover all usage cases (but if you see something odd, we can change or add warnings/errors).

fehiepsi commented on numpyro/contrib/nested_sampling.py:
```diff
@@ -0,0 +1,332 @@
+        # so this will raise an NotImplementedError for now.
+        # We will need scipy.special.gammaincinv, which is not available yet in JAX
+        # see issue: https://github.com/google/jax/issues/5350
+        gammas = uniform_reparam_transform(gamma_dist)(q)
```
TODO: wrap jaxns GammaPrior transform

fehiepsi commented on numpyro/contrib/nested_sampling.py:
```diff
@@ -0,0 +1,332 @@
+import jax.numpy as jnp
+
+try:
+    # jaxns uses x64 by default so we need to monkeypatch it here
```
monkeypatch -> undo jaxns changes

fehiepsi commented on numpyro/contrib/nested_sampling.py:
```diff
@@ -0,0 +1,332 @@
+    def diagnostics(self):
+        """
+        Plot diagnostics of the result. This is a wrapper of ``jaxns.plotting.plot_diagnostics``
+        and ``jaxns.plotting.plot_cornerplot``.
```
intersphinx jaxns (TODO: raise issue upstream)

fehiepsi commented on numpyro/contrib/nested_sampling.py:
```diff
@@ -0,0 +1,332 @@
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore", message=".*will be truncated to dtype float32.*"
+            )
```
raise issues in jaxns

fritzo reviewed (approved):
LGTM.   I assume `.cdf()` and `.icdf()` are covered by existing tests.

fehiepsi commented:
Yes, we have tests for both `cdf` and `icdf`. Thanks for reviewing, @fritzo!
